File: nim/triangle/triangle.nim
import sets, math

proc inequalityCheck(sides: array[3, int]): bool =
  if sides.sum == 0:
    return false
  else:
    sides[0] + sides[1] >= sides[2] and
    sides[1] + sides[2] >= sides[0] and
    sides[0] + sides[2] >= sides[1]

proc isEquilateral*(sides: array[3, int]): bool =
  sides.inequalityCheck and sides.toHashSet.len == 1

proc isIsosceles*(sides: array[3, int]): bool =
  sides.inequalityCheck and sides.toHashSet.len <= 2

proc isScalene*(sides: array[3, int]): bool =
  sides.inequalityCheck and sides.toHashSet.len == 3

File: ocaml/triangle/triangle.ml
let is_tri a b c =
  match (a, b, c) with
  | 0, _, _ | _, 0, _ | _, _, 0 -> false
  | _, _, _ -> a + b >= c && b + c >= a && a + c >= b

let is_equilateral a b c = is_tri a b c && a = b && b = c
let is_isosceles a b c = is_tri a b c && (a = b || a = c || b = c)
let is_scalene a b c = is_tri a b c && a <> b && a <> c && b <> c
